cap-ci/pipeline.template

```diff
@@ -67,6 +69,46 @@ resources:
 
 {{ $pipelineName := .Env.PIPELINE }}
 
+{{- $create_k8s := `
+            # create k8s cluster
+            export random_variable=$(hexdump -n 8 -e '2/4 "%08x"' /dev/urandom)
+            if [[ ${BACKEND} == "gke" ]]; then
+              printf "%s" '((gke-key-json))' > $PWD/gke-key.json
+              export GKE_CRED_JSON=$PWD/gke-key.json
+              export GKE_CLUSTER_NAME=ci-${BACKEND}-${random_variable}
+              export CLUSTER_NAME=${GKE_CLUSTER_NAME}
+              (cd catapult; make BACKEND=$BACKEND GKE_CREDS_JSON=$GKE_CRED_JSON GKE_CLUSTER_NAME=${GKE_CLUSTER_NAME} CLUSTER_NAME=${GKE_CLUSTER_NAME})
+              echo ${GKE_CLUSTER_NAME} > kubeconfig-pool/name
+              cp catapult/build${GKE_CLUSTER_NAME}/kubeconfig kubeconfig-pool/metadata
+            else
+              echo "CI cannot create ${BACKEND} cluster"
+              exit 1
+            fi
+` }}
+
+{{- $delete_k8s := `
+                  # delete k8s cluster
+                  if [[ ${BACKEND} == "gke" ]]; then
+                      echo "ls"
+                      ls
+                      pwd
+                      which gcloud
+                      printf "%s" '((gke-key-json))' > $PWD/gke-key.json
+                      export GKE_CRED_JSON=$PWD/gke-key.json
+                      export CLUSTER_NAME=$(cat pool.kube-hosts/name)
+                      pool_file=${pool_file:-pool.kube-hosts/metadata}
+                      export GKE_CLUSTER_ZONE=$(cat ${pool_file} | grep current-context | cut -d'_' -f3)
```

Comment:
few points here: 

- `terraform destroy` does not guarantee clean up.
- pools should not be used for multiple things, as it makes things difficult to understand by other.
- the correct way to handle terraform via concourse is using resources but we have decided against it, in favour catapult (plus it won't give us much benefit as these are ephemeral clusters)
- I think we can handle cleaning bug fix in catapult separately and use that once it is fixed. Otherwise, this ticket is just getting out of scope. A user is creating public clusters manually, so they can make sure things get cleaned after use. Making this a lower priority bug in catapult. But for CI, this is a blocker as all things needs to get cleaned up after use.

Reply:
> few points here:
> 
>  * `terraform destroy` does not guarantee clean up.

It does; you have deployed with terraform, if you don't clean with the terraform state file you will miss things sooner or later. Please don't deploy with terraform and clean manually with gcloud targetting a cluster_name which is not guaranteed to be a UID.

>  * pools should not be used for multiple things, as it makes things difficult to understand by others.

Do you honestly think that having 3 folders in the `gke-kube-hosts` pool: `clean`, `dirty`, `deployed`, 


>  * the correct way to handle terraform via concourse is using resources but we have decided against it, in favour catapult

Concourse best practices are geared for small projects, normally a monorepo of 1 language, where there's no need to cobble together several systems. Maybe blindly following Concourse suggestions when creating a distribution to be run on top of K8s is not a great idea.

Yes, we have all decided to use Catapult. Then, let's use it please. Don't make an unmaintainable pipeline that leaks secrets by calling gcloud directly, and destroys clusters in a wrong way.

> (plus it won't give us much benefit as these are ephemeral clusters)

Doing the things correctly will give you an implementation that works for cattle (public clouds) and pet (CaaSP4) clusters alike, with just needing to change foo in `BACKEND=foo`.

All that you are trying to wrongly redo has already been done, and doesn't need to be tied to Concourse.

>  * I think we can handle cleaning bug fix in catapult separately and use that once it is fixed. Otherwise, this ticket is just getting out of scope. A user is creating public clusters manually, so they can make sure things get cleaned after use. Making this a lower priority bug in catapult. But for CI, this is a blocker as all things needs to get cleaned up after use.

There's no cleaning bug in Catapult for GKE. Just do `BACKEND=gke make clean`. Just, use it.

Reply:
BACKEND=gke make clean, does not clean gke.

Reply:
I tested `terraform destroy` ( i.e `BACKEND=gke make clean`) today, to see if the issue persists.
It still leaves behind `disks` (Maybe because those disks are created on the fly by kubecf deployment and terraform has no record of it, but thats a different story.)

Reply:
Talked with Prabal about this today; cap-ci should indeed be using catapult as its interface to both backend and kubecf/cap interaction. Also discussed how we need an approach (in catapult) to generate the kubeconfig from a GKE cluster information (as IIRC GKE kubeconfig files are set to expire on ~24 hours or so). In cf-ci we would use a file that looked something like the following instead of a kubeconfig as our lockfiles for GKE clusters:

```
---
kind: ClusterReference
platform: gke
cluster-name: concourse-gke-4a8f6df9d0fbb62b
cluster-zone: us-west1-a
```

Since the pipeline has the gcloud cli and appropriate credentials, all it needs is the name and zone of the cluster in order to use `gcloud auth` to receive a new kubeconfig which won't expire for some time. This will allow us to have potentially longer-running GKE clusters as well as not have to worry about when the kubeconfig expires (which CI tasks shouldn't have to worry about)

Reply:
Additionally, terraform destroy indeed leaves disk resources behind, so it can't be relied on exclusively to clean up after itself

Reply:
That's a good find, thank you both for checking that.

Could you expand on the disks issue? I think it would indeed we nice to have a catapult or cap-terraform issue for it, and treat it there.
We could call `gcloud -q compute disks delete` just before calling [terraform destroy](https://github.com/SUSE/catapult/blob/c30fbb74d0d0cdba468a35ee13d82030a8b94c59/backend/gke/clean.sh#L16).

And also we should indeed not get kubeconfigs that have expiration time.
I would either fix cap-terraform or catapult for that ([here](https://github.com/SUSE/catapult/blob/master/backend/gke/kubeconfig.sh)).
But, are we sure that they actually expire? I remember this being changed in cap-terraform, and I have had GKE clusters running locally for a week without the need to reissue a new kubeconfig.